include/native/pipeline.h

```diff
@@ -21,398 +21,514 @@
 #ifndef GRPPI_NATIVE_PIPELINE_H
 #define GRPPI_NATIVE_PIPELINE_H
 
+#include "common/pack_traits.h"
+
 #include <experimental/optional>
 
 #include <thread>
 
 
 namespace grppi{
 
-template <typename InStream, typename OutStream, int currentStage, typename ...Stages>
- typename std::enable_if<(currentStage == (sizeof...(Stages)-1)), void>::type composed_pipeline(InStream& qin, pipeline_info<parallel_execution_native, Stages...> const & pipe, OutStream &qout,std::vector<std::thread> & tasks)
+template <typename InQueue, typename OutQueue, int Index, 
+          typename ... MoreTransformers,
+          internal::requires_index_last<Index,MoreTransformers...> = 0>
+void composed_pipeline(InQueue & input_queue, 
+                       const pipeline_info<parallel_execution_native, MoreTransformers...> & pipe, 
+                       OutQueue & output_queue, std::vector<std::thread> & tasks)
 {
-      composed_pipeline(pipe.exectype, qin, std::get<currentStage>(pipe.stages), qout, tasks);
+  composed_pipeline(pipe.exectype, input_queue, 
+      std::get<Index>(pipe.stages), output_queue, tasks);
 }
 
 
-template <typename InStream, typename OutStream, int currentStage, typename ...Stages>
- typename std::enable_if<(currentStage < (sizeof...(Stages)-1)), void>::type composed_pipeline(InStream& qin, pipeline_info<parallel_execution_native, Stages...> const & pipe, OutStream & qout,std::vector<std::thread> & tasks)
+template <typename InQueue, typename OutQueue, int Index, 
+          typename ... MoreTransformers,
+          internal::requires_index_not_last<Index,MoreTransformers...> = 0>
+void composed_pipeline(InQueue & input_queue, 
+                       const pipeline_info<parallel_execution_native,MoreTransformers...> & pipeline_obj, 
+                       OutQueue & output_queue, std::vector<std::thread> & tasks)
 {
-      typedef typename std::tuple_element<currentStage, decltype(pipe.stages)>::type lambdaType;
-//      typedef typename std::remove_reference<decltype(lambdaPointerType())>::type  lambdaType; 
-      typedef typename std::result_of< lambdaType (typename InStream::value_type::value_type) > ::type queueType;
-
-
-
-      static mpmc_queue<std::experimental::optional<queueType>> queueOut(pipe.exectype.queue_size,pipe.exectype.lockfree); 
-
-      composed_pipeline(pipe.exectype, qin, std::get<currentStage>(pipe.stages), queueOut, tasks);
-      composed_pipeline<mpmc_queue<std::experimental::optional<queueType>>,OutStream, currentStage+1, Stages ...>(queueOut,pipe,qout,tasks);
-       
+  using stage_type = 
+      typename std::tuple_element<Index,decltype(pipeline_obj.stages)>::type;
+  using input_type = typename InQueue::value_type;
+  using input_value_type = typename input_type::value_type;
+  using result_type = 
+      typename std::result_of<stage_type(input_type)>::type;
+
+  // TODO: Why static?
```

Comment:
Problems I see:
1. Thread safety. Is a race condition possible?
2. What if I have multiple pipelines at the same time. Could it be possible to have inconsistencies?